File: src/main/java/com/jaffa/rpc/lib/request/Request.java
package com.jaffa.rpc.lib.request;

import com.jaffa.rpc.lib.callbacks.Callback;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.TimeUnit;

public interface Request<T> {
    Request<T> withTimeout(long timeout, @NotNull TimeUnit unit);

    T executeSync();

    Request<T> onModule(@NotNull String moduleId);

    void executeAsync(@NotNull String key, @NotNull Class<? extends Callback<T>> listener);
}

File: src/main/kotlin/com/jaffa/rpc/lib/request/Request.kt
package com.jaffa.rpc.lib.request

import com.jaffa.rpc.lib.callbacks.Callback
import java.util.concurrent.TimeUnit

interface Request<T> {

    fun withTimeout(timeout: Long?, unit: TimeUnit?): Request<T>

    fun executeSync(): T

    fun onModule(moduleId: String?): Request<T>

    fun executeAsync(key: String?, listener: Class<out Callback<T>>?)

}